<a href="https://colab.research.google.com/github/KruglikovAnton/megafons_clients_investigation/blob/main/megafon_investigation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns
import numpy as np

![](https://sun9-32.userapi.com/C6vN5pHQW7hlseIt0AtTN7qNDQbDnHXdt1P2NA/ILntXqa-RrA.jpg)

## Codebook

`megafon.csv` содержит следующие значения: <br><br>
&nbsp;&nbsp;&nbsp;&nbsp; `user_id` — идентификатор абонента;<br>
&nbsp;&nbsp;&nbsp;&nbsp; `Q1` — ответ на первый вопрос;<br>
&nbsp;&nbsp;&nbsp;&nbsp; `Q2` — ответ на второй вопрос;<br>
&nbsp;&nbsp;&nbsp;&nbsp; `Total Traffic(MB)` — объем трафика передачи данных <sup>1 </sup>; <br>
&nbsp;&nbsp;&nbsp;&nbsp; `Downlink Throughput(Kbps)` — средняя скорость «к абоненту» <sup>2 </sup>;<br>
&nbsp;&nbsp;&nbsp;&nbsp; `Uplink Throughput(Kbps)`— средняя скорость «от абонента» <sup>3 </sup>;<br>
&nbsp;&nbsp;&nbsp;&nbsp; `Downlink TCP Retransmission Rate(%)` — частота переотправок пакетов «к абоненту» <sup>4 </sup>;<br>
&nbsp;&nbsp;&nbsp;&nbsp; `Video Streaming Download Throughput(Kbps)` — скорость загрузки потокового видео <sup>5 </sup>;<br>
&nbsp;&nbsp;&nbsp;&nbsp; `Video Streaming xKB Start Delay(ms)` — задержка старта воспроизведения видео <sup>6 </sup>;<br>
&nbsp;&nbsp;&nbsp;&nbsp; `Web Page Download Throughput(Kbps)` — скорость загрузки web-страниц через браузер <sup>7 </sup>;<br>
&nbsp;&nbsp;&nbsp;&nbsp; `Web Average TCP RTT(ms)` — пинг при просмотре web-страниц<sup>8 </sup>.<br>


<sup>1 </sup> — Насколько активно абонент использует мобильный интернет.<br>
<sup>2 </sup> — Считается по всему трафику передачи данных.<br>
<sup>3 </sup> — Считается по всему трафику передачи данных.<br>
<sup>4 </sup> — Чем выше, тем хуже. Если в канале возникает ошибка, пакет переотправляется. Снижается полезная скорость.<br>
<sup>5 </sup> — Чем выше, тем лучше — меньше прерываний и лучше качество картинки.<br>
<sup>6 </sup> — Сколько времени пройдёт между нажатием на кнопку Play и началом воспроизведения видео. Чем меньше это время, тем быстрее начинается воспроизведение.<br>
<sup>7 </sup> — Чем выше, тем лучше.<br>
<sup>8 </sup> — Чем меньше, тем лучше — быстрее загружаются web-страницы.<br>

Первый технический показатель представлен как сумма за период в одну неделю перед участием в опросе. Остальные технические показатели отображают среднее значение по данному признаку за период в одну неделю перед участием в опросе.

In [ ]:
start_df = pd.read_csv('megafon.csv')
start_df.head(5)

user_id Q1       Q2  Total Traffic(MB)  Downlink Throughput(Kbps)  \
0        1  5      NaN          775.48846                     360.13   
1        2  5        4          861.96324                    3023.54   
2        3  1        4          261.11860                     790.96   
3        4  8        3          179.18564                    2590.97   
4        5  2  2, 3, 4          351.99208                     731.61   

   Uplink Throughput(Kbps)  Downlink TCP Retransmission Rate(%)  \
0                    86.56                                 3.93   
1                   411.18                                 1.27   
2                    34.20                                 1.79   
3                   325.88                                 0.80   
4                   223.54                                 1.15   

   Video Streaming Download Throughput(Kbps)  \
0                                    1859.15   
1                                     667.47   
2                                    1079.60   
3                                    7053.81   
4                                    4550.38   

   Video Streaming xKB Start Delay(ms)  Web Page Download Throughput(Kbps)  \
0                                 2309                             1007.82   
1                                 2080                              255.36   
2                                 6367                              535.85   
3                                 3218                             1221.02   
4                                 1767                             2336.56   

   Web Average TCP RTT(ms)  
0                       83  
1                      425  
2                      485  
3                       51  
4                       68

In [ ]:
start_df.describe(include='all')

user_id    Q1    Q2  Total Traffic(MB)  Downlink Throughput(Kbps)  \
count   3112.000000  3110  1315        3112.000000                3112.000000   
unique          NaN    49    51                NaN                        NaN   
top             NaN    10     3                NaN                        NaN   
freq            NaN   846   190                NaN                        NaN   
mean    1556.500000   NaN   NaN         406.282339                2220.774023   
std      898.501345   NaN   NaN         220.593815                2629.824405   
min        1.000000   NaN   NaN         100.026680                  13.260000   
25%      778.750000   NaN   NaN         214.261192                 652.297500   
50%     1556.500000   NaN   NaN         363.999720                1404.230000   
75%     2334.250000   NaN   NaN         572.524325                2823.645000   
max     3112.000000   NaN   NaN         887.078270               39817.640000   

        Uplink Throughput(Kbps)  Downlink TCP Retransmission Rate(%)  \
count               3112.000000                          3112.000000   
unique                      NaN                                  NaN   
top                         NaN                                  NaN   
freq                        NaN                                  NaN   
mean                 170.702381                             1.834518   
std                  204.031286                             2.074593   
min                    6.110000                             0.050000   
25%                   70.735000                             0.720000   
50%                  119.435000                             1.300000   
75%                  210.617500                             2.240000   
max                 4141.850000                            31.200000   

        Video Streaming Download Throughput(Kbps)  \
count                                 3112.000000   
unique                                        NaN   
top                                           NaN   
freq                                          NaN   
mean                                  5728.806848   
std                                   5161.475235   
min                                      0.000000   
25%                                   2324.550000   
50%                                   4436.500000   
75%                                   7611.440000   
max                                  58566.030000   

        Video Streaming xKB Start Delay(ms)  \
count                           3112.000000   
unique                                  NaN   
top                                     NaN   
freq                                    NaN   
mean                            2107.596080   
std                             1582.423057   
min                              239.000000   
25%                             1280.000000   
50%                             1672.000000   
75%                             2312.000000   
max                            21392.000000   

        Web Page Download Throughput(Kbps)  Web Average TCP RTT(ms)  
count                          3112.000000              3112.000000  
unique                                 NaN                      NaN  
top                                    NaN                      NaN  
freq                                   NaN                      NaN  
mean                           2100.258901               194.195051  
std                            2062.509549               267.387914  
min                               0.000000                 0.000000  
25%                            1018.002500                74.000000  
50%                            1670.695000               118.000000  
75%                            2519.987500               207.000000  
max                           35603.970000              3471.000000

В Q1 и Q2 49 и 51 уникальных значения из общей выборки в 3110 и 1315 соответственно, что связано с своим вариантом ответа или nan.

In [ ]:
#Отсортируем по значению Q1
start_df.sort_values('Q1')

user_id                                                 Q1   Q2  \
491       492                                       ***** ** ***  NaN   
158       159                                                  0  NaN   
1133     1134                                                  0  NaN   
1991     1992                                                  0  NaN   
117       118                                                  0  NaN   
...       ...                                                ...  ...   
834       835  Чем даль ше,тем лучше.Спасибо за ваш труд.Оцен...  NaN   
3041     3042  Я в Смол. Области живу сейчас, не пользуюсь те...  NaN   
1365     1366                      Я ценой услуг не удовлетворен  NaN   
371       372                                                NaN  NaN   
3026     3027                                                NaN  NaN   

      Total Traffic(MB)  Downlink Throughput(Kbps)  Uplink Throughput(Kbps)  \
491           479.50838                    8675.19                   230.06   
158           126.51761                     361.56                    46.89   
1133          359.66875                    3578.95                    58.31   
1991          140.70227                    3575.30                   150.95   
117           837.55274                    2511.24                   113.10   
...                 ...                        ...                      ...   
834           886.05369                    1115.41                   937.26   
3041          338.04205                     167.67                    57.20   
1365          543.14687                    4818.84                    81.10   
371           286.62191                    2320.25                    75.60   
3026          875.62280                    7258.66                   226.08   

      Downlink TCP Retransmission Rate(%)  \
491                                  0.33   
158                                  0.82   
1133                                 2.34   
1991                                 0.83   
117                                  0.39   
...                                   ...   
834                                  1.20   
3041                                 1.16   
1365                                 0.72   
371                                  3.84   
3026                                 0.60   

      Video Streaming Download Throughput(Kbps)  \
491                                     6818.51   
158                                      965.21   
1133                                    5634.10   
1991                                    7929.53   
117                                     4976.77   
...                                         ...   
834                                     4439.31   
3041                                     363.19   
1365                                    6786.28   
371                                     4499.92   
3026                                    8336.42   

      Video Streaming xKB Start Delay(ms)  Web Page Download Throughput(Kbps)  \
491                                  1562                             2797.14   
158                                  3564                             1709.08   
1133                                 1066                             1745.89   
1991                                 1884                             3048.05   
117                                  2244                             1544.19   
...                                   ...                                 ...   
834                                  1930                              753.67   
3041                                  782                              166.81   
1365                                 1129                             2494.68   
371                                  1790                             1035.64   
3026                                 1087                             3112.83   

      Web Average TCP RTT(ms)  
491                  

Как видно из отсортированной таблицы существуют значения, где в Q1 и Q2 присутствуют nan, а также значения, в которых в Q1 0, а в Q2 nan. На данном этапе уберем подобные значения.

In [ ]:
main_df = start_df[start_df.Q1.isin(list(map(str, list(range(1,10)))))]
main_df.describe(include='all')

Как видим это уменьшило число данных на треть, что говорит о необходимости их исследования